### PR TITLE
Feature/remove side effects

### PR DESCRIPTION
**Pull request by TroyAlford**
There are currently a number of functions in-use by this lib which alter the native prototypes, such as `Array.prototype` and `Object.prototype`, which is generally considered an anti-pattern.

This PR does the following:
 - refactors each of these functions, and each of their uses to avoid prototype side-effects
 - removes one function which appears to be entirely unused (`glaem_merge`)

Among other things, this prevents an issue which occurs if, for some reason, `glaemscribe` is loaded twice.

### CODE REVIEW

**Comment by TroyAlford** [comment]
Thanks for building/maintaining this project, btw. :) I'm currently working on a wrapper to expose the functionality as both a react component and as web-components. The web-component implementation, in particular, will make it pretty trivial to use glaemscribe on various sites. I'm working out an API like:
```
<queta-text lang="quenya-tengwar-classical" font="tengwar_guni_sindarin">
  Ai ! laurië lantar lassi súrinen ,
  yéni únótimë ve rámar aldaron !
  Yéni ve lintë yuldar avánier
  mi oromardi lissë-miruvóreva
  Andúnë pella , Vardo tellumar
  nu luini yassen tintilar i eleni
  ómaryo airetári-lírinen .

  Sí man i yulma nin enquantuva ?
</queta-text>
```

File: build/web/glaemscribe/js/charsets/eldamar.cst.js [hunk 1]
```diff
@@ -0,0 +1 @@
... THESE **\\\n\\char e058  SARINCE_ENDING_LONG SARINCE_FLOURISHED\n\\char e059  SARINCE COMBINING_SARINCE SHOOK_RIGHT_L SHOOK_LEFT_L \n\\char e05A  ?\n\\char e05B  ?\n\n\\char e065  PUNCT_EXCLAM\n\\char e066  PUNCT_INTERR\n\\char e067  ELVISH_PAREN PUNCT_PAREN_L PUNCT_PAREN_R PUNCT_PAREN_L_ALT PUNCT_PAREN_R_ALT BOOKMARK_SIGN\n\\char e068  PUNCT_TILD\n \n\\char e069  PUNCT_DTILD RING_MARK_L RING_MARK_R\n\n\\char e06A  DQUOT_OPEN\n\\char e06B  DQUOT_CLOSE\n\\char e06C  THORIN_EXCLAMATION\n\n\\char e070 NUM_0 \n\\char e071 NUM_1 \n\\char e072 NUM_2 \n\\char e073 NUM_3 \n\\char e074 NUM_4 \n\\char e075 NUM_5 \n\\char e076 NUM_6 \n\\char e077 NUM_7 \n\\char e078 NUM_8 \n\\char e079 NUM_9 \n\\char e07A NUM_10\n\\char e07B NUM_11\n\\char e07C NUM_12\n\n\\char e07D CIRC_TEHTA_INF\n\n\\** Ligatured results are mapped here **\\\n\\**\n\\char 10037 AHA_TINCO\n\\char 10038 HWESTA_TINCO\n\\char 10039 ANCA_CLOSED SILME_AHA\n**\\\n\n\\char 204A OLD_ENGLISH_AND\n\n\\char 200d ZWJ\n\n\\** Ligatured sequences used as one entity in modes **\\\n\\beg seq AHA_TINCO\n  AHA ZWJ TINCO\n\\end\n\n\\beg seq HWESTA_TINCO\n  HWESTA ZWJ TINCO\n\\end\n\n\\beg seq ANCA_CLOSED SILME_AHA\n  SILME ZWJ AHA\n\\end\n\n"
```
Comment (TroyAlford) [inline build/web/glaemscribe/js/charsets/eldamar.cst.js:1]: Not sure what caused this file to be generated. It appears to be new. Maybe I ran the build incorrectly? I ran `ruby ./bin_dev/make_all.rb`

**Comment by TroyAlford** [comment]
It appears that this change doesn't actually work for some reason... I'm now getting an error of: `"Cannot understand constant HYPHEN_WORD_JOINER"` - which led me to notice that the latest NPM publish was Aug 8, 2019 - and there's a lot of change between then and now.

Is it possible the JS build has become broken based on those changes?

**Comment by BenTalagan** [comment]
Hi Troy, thanks a lot for your feedback, and for the PR! I do 100% agree with the original argument. This is bad practice especially in advanced projects that load multiple libs and this may interfere with other components. I'll merge this ASAP in the master after review. The idea of extending Array is probably coming from the fact that I had ported the library from ruby to JS having both source standing side by side, and tried to keep the JS syntax as close as possible to the ruby code, hence the `.each` (I just can't remember). But this really needs to be changed.

Also, as you've stated in one of your precedent email it's been a while since the last release. It's not that there hasn't been any work done this year on glaemscribe. We've been working hard with my linguist mate Bertrand on a decent english mode, and it's been very challenging, but we're not too far from the final milestone. Still, there's a lot of doc to be written and I was waiting for the english mode to be ready to pack a new release for glaemscribe. But it may still take a little while (probably a few weeks/months) since we're both busy atm, and things always get longer than expected.

So what I'll probably do is pack a minor intermediary release to have your changes in prod :-)

**Review by BenTalagan — APPROVED** [review]

**Comment by TroyAlford** [comment]
Much appreciated! I'd be happy to contribute more, btw, if its welcome.

**Pull request merged**